README.md
The fundamental package for financial computing with Dart.

Not in pub.dev yet. This fork works with dart 3.
This package provides a set of high-level functions and classes that make it easy to compute financial ratios, patterned after spreadsheet computations. It's platform-independent, supports iOS and Android.

# Using

The easiest way to use this library is via the top-level functions. They allow you to make financial calculations with minimal hassle:

## Future Value calculation

What is the future value after 30 years of saving $1,000 now, with an additional monthly savings of $100. Assume the interest rate is 7% (annually) compounded monthly?

```dart
  print(Finance.fv(rate: 0.07 / 12, nper: 30 * 12, pmt: -100, pv: -1000));
```

Thus, saving $1,000 today and then $100 a month at 7% annual interest leads to $130,113.59 available to spend in 30 years from now.

## Payments calculation

What is the monthly payment needed to pay off a $100,000 loan in 10 years at an annual interest rate of 2.5%?

```dart
  print(Finance.pmt(rate: 0.025 / 12, nper: 10 * 12, pv: 100000));
```  

Thus, in order to pay-off (i.e., have a future-value of 0) the $100,000 obtained today, a monthly payment of $942.69 would be required.  Note that this example illustrates usage of future value having a default value of $0.

## Number of period calculation

If we only had $500/month to pay towards the loan, how long would it take to pay-off a loan of $100,000 at 3% annual interest?
 
 ```dart
   print(Finance.nper(rate: 0.03 / 12, pmt: -500, pv: 10000));
 ```  
 
So, over 20 months would be required to pay off the loan.

## Principal and Interest payment calculation

What is the amortization schedule for a 1 year loan of $5,000 at 10% interest per year compounded monthly? 
And the total interest payments?

```dart
  final Iterable<Map<String, num>> payments =
      List<int>.generate(12, (int index) => index + 1).map((int per) =>
          <String, num>{
            'per': per,
            'pmt': Finance.pmt(rate: 0.1 / 12, nper: 1 * 12, pv: 5000),
            'ppmt':
                Finance.ppmt(rate: 0.1 / 12, per: per, nper: 1 * 12, pv: 5000),
            'ipmt':
                Finance.ipmt(rate: 0.1 / 12, per: per, nper: 1 * 12, pv: 5000),
          });

  payments.forEach(print);

  final num interestPaid =
      payments.fold(0, (num p, Map<String, num> c) => p + c['ipmt']);
  print(interestPaid);
 ```
 
Thus, the total interest paid are close to $275.

## Present Value calculation

What is the present value (e.g., the initial investment) of an investment that needs to total $20,000 
after 10 years of saving $100 every month?  Assume the interest rate is 7% (annually) compounded monthly.

```dart
  print(Finance.pv(rate: 0.05 / 12, nper: 10 * 12, pmt: -100, fv: 15692.93));
```

So, to end up with $20,000 in 10 years saving $100 a month at 7% annual interest, one's initial deposit 
should be $1,339.28.

## Interest rate calculation

What is the interest rate for a $8,000 loan if the loan term is 5 years and and payments are $152.50 monthly?

```dart
  print(Finance.rate(nper: 60, pmt: -152.5, pv: 8000, fv: 0));
```

The calculated interest rate is 0.45% monthly, or 5.42% annually.

## Net present value calculation

What is the Net Present Value of an investment that requires $15,000 deposit now and then makes the following 
withdrawals at regular (fixed) intervals:  $1,500, $2,500, $3,500, $4,500, $6,000?.  Assuming the ending value is $0
and a discount rate of 5%.

```dart
  print(Finance.npv(rate: 0.05, values: <num>[-15000, 1500, 2500, 3500, 4500, 6000]));
```

So, this investment is worth $122.894 today.

## Internal Rate of Return calculation

What is the Internal Rate of return of an investment that requires $15,000 deposit now and then makes the 
following withdrawals at regular (fixed) intervals:  $1,500, $2,500, $3,500, $4,500, $6,000?.  Assuming the ending 
value is $0.

```dart
  print(Finance.irr(rate: 0.05, values: <num>[-15000, 1500, 2500, 3500, 4500, 6000]));
```

So, the IRR of this investment is 5.24%.


CHANGELOG.md
## [0.1.2] - 6/21/23

*  Forked - refactored, upgraded.

## [0.1.0] - 3/9/19

*  Added nper, ppmt, ipmt, pv, rate, npv and irr. Stabilize interfaces

## [0.0.5] - 2/9/19

*  Extracted quote retrieval to finance_quote package. Added fv and pmt functions.

## [0.0.4] - 1/9/19

*  Documentation fixes.

## [0.0.3] - 1/9/19

*  Documentation fixes.

## [0.0.2] - 1/9/19

*  Documentation fixes.

## [0.0.1] - 1/9/19

*  Initial implementation of finance plugin.



lib/finance.dart
// Copyright 2019 Ismael Jiménez. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

library finance;

import 'dart:math';

class Finance {
  /// Returns the value at the end of the `nper` periods.
  ///
  /// The `rate` argument specifies the rate of interest as decimal (not per cent) per period.
  ///
  /// The `nper` argument specifies the number of compounding periods.
  ///
  ///  The `pmt` argument specifies the payment. By convention, the negative sign represents
  ///  cash flow out (i.e. money not available today).
  ///
  ///  The `pv` argument specifies the present value. By convention, the negative sign represents
  ///  cash flow out (i.e. money not available today).
  ///
  /// If specified, the `end` argument specifies when payments are due, at the end or beginning
  /// of each period. Per default, end of each period is selected.
  ///
  ///  Notes
  ///  -----
  ///    The future value is computed by solving the equation::
  ///     fv +
  ///     pv*(1+rate)**nper +
  ///     pmt*(1 + rate*end)/rate*((1 + rate)**nper - 1) == 0
  ///    or, when ``rate == 0``::
  ///     fv + pv + pmt * nper == 0
  ///
  ///  References
  ///  ----------
  ///     [WRW] Wheeler, D. A., E. Rathke, and R. Weir (Eds.) (2009, May).
  ///     Open Document Format for Office Applications (OpenDocument)v1.2,
  ///     Part 2: Recalculated Formula (OpenFormula) Format - Annotated Version,
  ///     Pre-Draft 12. Organization for the Advancement of Structured Information
  ///     Standards (OASIS). Billerica, MA, USA. [ODT Document].
  ///     Available:
  ///     http://www.oasis-open.org/committees/documents.php?wg_abbrev=office-formula
  ///     OpenDocument-formula-20090508.odt
  static double fv(
      {required double rate,
      required double nper,
      required double pmt,
      required double pv,
      bool end = true}) {
    final int when = end ? 0 : 1;
    final double temp = pow(1 + rate, nper).toDouble();
    final double fact =
        (rate == 0) ? nper : ((1 + rate * when) * (temp - 1) / rate);
    return -(pv * temp + pmt * fact);
  }

  /// Returns the payment against loan principal plus interest.  By convention, the negative sign represents
  /// cash flow out.
  ///
  /// The `rate` argument specifies the rate of interest as decimal (not per cent) per period.
  ///
  /// The `nper` argument specifies the number of compounding periods.
  ///
  ///  The `pv` argument specifies the present value. By convention, the negative sign represents
  ///  cash flow out.
  ///
  /// If specified, the `fv` argument specifies the future value. (default=0). By convention, the negative
  /// sign represent cash flow out.
  ///
  /// If specified, the `end` argument specifies when payments are due, at the end or beginning
  /// of each period. (default=true).
  ///
  ///  Notes
  ///  -----
  ///    The payment is computed by solving the equation::
  ///     fv +
  ///     pv*(1 + rate)**nper +
  ///     pmt*(1 + rate*end)/rate*((1 + rate)**nper - 1) == 0
  ///    or, when ``rate == 0``::
  ///      fv + pv + pmt * nper == 0
  ///    for ``pmt``.
  ///    Note that computing a monthly mortgage payment is only
  ///    one use for this function.  For example, pmt returns the
  ///    periodic deposit one must make to achieve a specified
  ///    future balance given an initial deposit, a fixed,
  ///    periodically compounded interest rate, and the total
  ///    number of periods.
  ///
  ///  References
  ///  ----------
  ///     [WRW] Wheeler, D. A., E. Rathke, and R. Weir (Eds.) (2009, May).
  ///     Open Document Format for Office Applications (OpenDocument)v1.2,
  ///     Part 2: Recalculated Formula (OpenFormula) Format - Annotated Version,
  ///     Pre-Draft 12. Organization for the Advancement of Structured Information
  ///     Standards (OASIS). Billerica, MA, USA. [ODT Document].
  ///     Available:
  ///     http://www.oasis-open.org/committees/documents.php
  ///     ?wg_abbrev=office-formulaOpenDocument-formula-20090508.odt
  static double pmt(
      {required double rate,
      required double nper,
      required double pv,
      double fv = 0,
      bool end = true}) {
    final int when = end ? 0 : 1;
    final double temp = pow(1 + rate, nper).toDouble();
    final double maskedRate = (rate == 0) ? 1 : rate;
    final double fact = (rate == 0)
        ? nper
        : ((1 + maskedRate * when) * (temp - 1) / maskedRate);
    return -(fv + pv * temp) / fact;
  }

  /// Returns the number of periodic payments.
  ///
  /// The `rate` argument specifies the rate of interest as decimal (not per cent) per period.
  ///
  ///  The `pmt` argument specifies the payment. By convention, the negative sign represents
  ///  cash flow out (i.e. money not available today).
  ///
  ///  The `pv` argument specifies the present value. By convention, the negative sign represents
  ///  cash flow out.
  ///
  /// If specified, the `fv` argument specifies the future value. (default=0). By convention, the negative
  /// sign represent cash flow out.
  ///
  /// If specified, the `end` argument specifies when payments are due, at the end or beginning
  /// of each period. (default=true).
  ///
  ///  Notes
  ///  -----
  ///    The number of periods ``nper`` is computed by solving the equation::
  ///     fv + pv*(1+rate)**nper + pmt*(1+rate*(end))/rate*((1+rate)**nper-1) = 0
  ///    but if ``rate = 0`` then::
  ///     fv + pv + pmt*nper = 0
  static double nper(
      {required double rate,
      required double pmt,
      required double pv,
      double fv = 0,
      bool end = true}) {
    final int when = end ? 0 : 1;

    try {
      final double A = -(fv + pv) / pmt;
      final double z = pmt * (1 + rate * when) / rate;
      final double B = log((-fv + z) / (pv + z)) / log(1 + rate);
      return (rate == 0) ? A : B;
    } catch (e) {
      return (-fv + pv) / pmt;
    }
  }

  /// Returns the interest portion of a payment.
  ///
  /// The `rate` argument specifies the rate of interest as decimal (not per cent) per period.
  ///
  ///  The `per` argument specifies the interest paid against the loan changes during the life or the loan.. By convention, the negative sign represents
  ///  cash flow out (i.e. money not available today).
  ///
  /// The `nper` argument specifies the number of compounding periods.
  ///
  ///  The `pv` argument specifies the present value. By convention, the negative sign represents
  ///  cash flow out.
  ///
  /// If specified, the `fv` argument specifies the future value. (default=0). By convention, the negative
  /// sign represent cash flow out.
  ///
  /// If specified, the `end` argument specifies when payments are due, at the end or beginning
  /// of each period. (default=true).
  ///
  static double ipmt(
      {required double rate,
      required double per,
      required double nper,
      required double pv,
      double fv = 0,
      bool end = true}) {
    final double totalPmt = pmt(rate: rate, nper: nper, pv: pv, fv: fv, end: end);
    double ipmt =
        _rbl(rate: rate, per: per, pmt: totalPmt, pv: pv, end: end) * rate;
    ipmt = end ? ipmt : ipmt / (1 + rate);
    ipmt = (!end && (per == 1)) ? 0 : ipmt;
    return ipmt;
  }

  //  This function is here to simply have a different name for the 'fv'
  //  function to not interfere with the 'fv' keyword argument within the 'ipmt'
  //  function.  It is the 'remaining balance on loan'.
  static double _rbl(
      {required double rate,
      required double per,
      required double pmt,
      required double pv,
      bool end = true}) {
    return fv(rate: rate, nper: per - 1, pmt: pmt, pv: pv, end: end);
  }

  /// Returns the payment against loan principal.
  ///
  /// The `rate` argument specifies the rate of interest as decimal (not per cent) per period.
  ///
  ///  The `per` argument specifies the interest paid against the loan changes during the life or the loan.. By convention, the negative sign represents
  ///  cash flow out (i.e. money not available today).
  ///
  /// The `nper` argument specifies the number of compounding periods.
  ///
  ///  The `pv` argument specifies the present value. By convention, the negative sign represents
  ///  cash flow out.
  ///
  /// If specified, the `fv` argument specifies the future value. (default=0). By convention, the negative
  /// sign represent cash flow out.
  ///
  /// If specified, the `end` argument specifies when payments are due, at the end or beginning
  /// of each period. (default=true).
  ///
  static double ppmt(
      {required double rate,
      required double per,
      required double nper,
      required double pv,
      double fv = 0,
      bool end = true}) {
    final double total = pmt(rate: rate, nper: nper, pv: pv, fv: fv, end: end);
    return total -
        ipmt(rate: rate, per: per, nper: nper, pv: pv, fv: fv, end: end);
  }

  /// Returns the present value of a series of payments or investments.
  ///
  /// The `rate` argument specifies the rate of interest as decimal (not per cent) per period.
  ///
  /// The `nper` argument specifies the number of compounding periods.
  ///
  ///  The `pmt` argument specifies the payment. By convention, the negative sign represents
  ///  cash flow out (i.e. money not available today).
  ///
  ///  The `fv` argument specifies the future value. By convention, the negative sign represents
  ///  cash flow out (i.e. money not available today).
  ///
  /// If specified, the `end` argument specifies when payments are due, at the end or beginning
  /// of each period. Per default, end of each period is selected.
  ///
  ///  Notes
  ///  -----
  ///    The present value is computed by solving the equation::
  ///     fv +
  ///     pv*(1 + rate)**nper +
  ///     pmt*(1 + rate*end)/rate*((1 + rate)**nper - 1) = 0
  ///    or, when ``rate = 0``::
  ///     fv + pv + pmt * nper = 0
  ///    for `pv`, which is then returned.
  ///
  ///  References
  ///  ----------
  ///    [WRW] Wheeler, D. A., E. Rathke, and R. Weir (Eds.) (2009, May).
  ///    Open Document Format for Office Applications (OpenDocument)v1.2,
  ///    Part 2: Recalculated Formula (OpenFormula) Format - Annotated Version,
  ///    Pre-Draft 12. Organization for the Advancement of Structured Information
  ///    Standards (OASIS). Billerica, MA, USA. [ODT Document].
  ///    Available:
  ///    http://www.oasis-open.org/committees/documents.php?wg_abbrev=office-formula
  ///    OpenDocument-formula-20090508.odt
  static double pv(
      {required double rate,
      required double nper,
      required double pmt,
      required double fv,
      bool end = true}) {
    final int when = end ? 0 : 1;
    final double temp = pow(1 + rate, nper).toDouble();
    final double fact =
        (rate == 0) ? nper : ((1 + rate * when) * (temp - 1) / rate);
    return -(fv + pmt * fact) / temp;
  }

  // Computed with Sage
  //  (y + (r + 1)^n*x + p*((r + 1)^n - 1)*(r*w + 1)/r)/(n*(r + 1)^(n - 1)*x -
  //  p*((r + 1)^n - 1)*(r*w + 1)/r^2 + n*p*(r + 1)^(n - 1)*(r*w + 1)/r +
  //  p*((r + 1)^n - 1)*w/r)
  static double _g_div_gp(double r, double n, double p, double x, double y, double w) {
    final double t1 = pow(r + 1, n).toDouble();
    final double t2 = pow(r + 1, n - 1).toDouble();
    return (y + t1 * x + p * (t1 - 1) * (r * w + 1) / r) /
        (n * t2 * x -
            p * (t1 - 1) * (r * w + 1) / pow(r, 2) +
            n * p * t2 * (r * w + 1) / r +
            p * (t1 - 1) * w / r);
  }

  /// Returns the rate of interest per period.
  ///
  /// The `nper` argument specifies the number of compounding periods.
  ///
  ///  The `pmt` argument specifies the payment. By convention, the negative sign represents
  ///  cash flow out (i.e. money not available today).
  ///
  ///  The `pv` argument specifies the present value. By convention, the negative sign represents
  ///  cash flow out.
  ///
  ///  The `fv` argument specifies the future value. By convention, the negative sign represents
  ///  cash flow out (i.e. money not available today).
  ///
  /// If specified, the `end` argument specifies when payments are due, at the end or beginning
  /// of each period. Per default, end of each period is selected.
  ///
  /// If specified, the `guess` argument specifies the starting guess for solving the rate of interest.
  ///
  /// If specified, the `tol` argument specifies the required tolerance for the solution.
  ///
  /// If specified, the `maxIter` argument specifies the maximum iterations in finding the solution.
  ///
  ///  Notes
  ///  -----
  ///    The rate of interest is computed by iteratively solving the
  ///    (non-linear) equation::
  ///     fv + pv*(1+rate)**nper + pmt*(1+rate*when)/rate * ((1+rate)**nper - 1) = 0
  ///    for ``rate``.
  ///  -----
  ///    Use Newton's iteration until the change is less than 1e-6
  ///    for all values or a maximum of 100 iterations is reached.
  ///    Newton's rule is
  ///    r_{n+1} = r_{n} - g(r_n)/g'(r_n)
  ///     where
  ///    g(r) is the formula
  ///    g'(r) is the derivative with respect to r.
  ///
  ///  References
  ///  ----------
  ///    Wheeler, D. A., E. Rathke, and R. Weir (Eds.) (2009, May). Open Document
  ///    Format for Office Applications (OpenDocument)v1.2, Part 2: Recalculated
  ///    Formula (OpenFormula) Format - Annotated Version, Pre-Draft 12.
  ///    Organization for the Advancement of Structured Information Standards
  ///    (OASIS). Billerica, MA, USA. [ODT Document]. Available:
  ///    http://www.oasis-open.org/committees/documents.php?wg_abbrev=office-formula
  ///    OpenDocument-formula-20090508.odt
  static double rate(
      {required double nper,
      required double pmt,
      required double pv,
      required double fv,
      bool end = true,
      double guess = 0.1,
      double tol = 1e-6,
      double maxIter = 100}) {
    final int when = end ? 0 : 1;

    double rn = guess;
    double iterator = 0;
    bool close = false;
    while ((iterator < maxIter) && !close) {
      final double rnp1 = rn - _g_div_gp(rn, nper, pmt, pv, fv, when.toDouble());
      final double diff = (rnp1 - rn).abs();
      close = diff < tol;
      iterator += 1;
      rn = rnp1;
    }

    return rn;
  }

  /// Returns the NPV (Net Present Value) of a cash flow series.
  ///
  /// The `rate` argument specifies the discount rate.
  ///
  /// The `values` argument specifies the values of the time series of cash flows.
  /// The (fixed) time interval between cash flow "events" must be the same as that for
  /// which `rate` is given (i.e., if `rate` is per year, then precisely a year is understood
  /// to elapse between each cash flow event).  By convention, investments or "deposits" are
  /// negative, income or "withdrawals" are positive; `values` must begin with the initial
  /// investment, thus `values[0]` will typically be negative.
  ///
  /// Notes
  /// -----
  ///    Returns the result of: [G]_
  ///    .. math :: \\sum_{t=0}^{M-1}{\\frac{values_t}{(1+rate)^{t}}}
  ///
  /// References
  /// ----------
  ///    .. [G] L. J. Gitman, "Principles of Managerial Finance, Brief," 3rd ed.,
  ///       Addison-Wesley, 2003, pg. 346.
  static double npv({required double rate, required List<num> values}) {
    return List<int>.generate(values.length, (int index) => index)
        .map((int index) => values[index] / pow(1 + rate, index))
        .fold(0, (double p, double c) => p + c);
  }

  static double _npvPrime({required double rate, required List<num> values}) {
    return List<int>.generate(values.length, (int index) => index)
        .map((int index) => -index * values[index] / pow(1 + rate, index + 1))
        .fold(0, (double p, double c) => p + c);
  }

  static double _npv_div_npvPrime(double rate, List<num> values) {
    final double t1 = npv(rate: rate, values: values);
    final double t2 = _npvPrime(rate: rate, values: values);
    return t1 / t2;
  }

  /// Returns the Internal Rate of Return for periodic input values.
  ///
  /// The `values` argument specifies the values of the time series of cash flows.
  /// By convention, investments or "deposits" are negative, income or "withdrawals"
  /// are positive; `values` must begin with the initial investment, thus `values[0]`
  /// will typically be negative.
  ///
  /// If specified, the `guess` argument specifies the starting guess for solving the rate of interest.
  ///
  /// If specified, the `tol` argument specifies the required tolerance for the solution.
  ///
  /// If specified, the `maxIter` argument specifies the maximum iterations in finding the solution.
  ///
  /// Notes
  /// -----
  ///    The internal rate of interest is computed by iteratively solving the
  ///    (non-linear) equation::
  ///     npv(rate, values) = 0
  ///    for ``rate``.
  ///  -----
  ///    Use Newton's iteration until the change is less than 1e-6
  ///    for all values or a maximum of 100 iterations is reached.
  ///    Newton's rule is
  ///    r_{n+1} = r_{n} - g(r_n)/g'(r_n)
  ///     where
  ///    g(r) is the formula
  ///    g'(r) is the derivative with respect to r.

  static double irr(
      {required List<num> values,
      double guess = 0.1,
      double tol = 1e-6,
      double maxIter = 100}) {
    double rn = guess;
    double iterator = 0;
    bool close = false;
    while ((iterator < maxIter) && !close) {
      final double rnp1 = rn - _npv_div_npvPrime(rn, values);
      final double diff = (rnp1 - rn).abs();
      close = diff < tol;
      iterator += 1;
      rn = rnp1;
    }

    return rn;
  }
}


example/lib/main.dart
import 'package:finance/finance.dart';

void main(List<String> arguments) {
  // What is the future value after 30 years of saving $1000 now, with
  // an additional monthly savings of $100.  Assume the interest rate is
  // 7% (annually) compounded monthly?
  print(Finance.fv(rate: 0.07 / 12, nper: 30 * 12, pmt: -100, pv: -1000));

  // What is the monthly payment needed to pay off a $100,000 loan in 10
  // years at an annual interest rate of 2.5%?
  print(Finance.pmt(rate: 0.025 / 12, nper: 10 * 12, pv: 100000));

  // If you only had $500/month to pay towards the loan, how long would it take
  // to pay-off a loan of $10,000 at 3% annual interest?
  print(Finance.nper(rate: 0.03 / 12, pmt: -500, pv: 10000));

  // What is the amortization schedule for a 1 year loan of $5000 at 10%
  // interest per year compounded monthly?
  final Iterable<Map<String, num>> payments =
      List<int>.generate(12, (int index) => index + 1).map((int per) =>
          <String, num>{
            'per': per,
            'pmt': Finance.pmt(rate: 0.1 / 12, nper: 1 * 12, pv: 5000),
            'ppmt':
                Finance.ppmt(rate: 0.1 / 12, per: per.toDouble(), nper: 1 * 12, pv: 5000),
            'ipmt':
                Finance.ipmt(rate: 0.1 / 12, per: per.toDouble(), nper: 1 * 12, pv: 5000),
          });

  payments.forEach(print);
  final num interestPaid =
      payments.fold(0, (num p, Map<String, num> c) => p + c['ipmt']!);
  print(interestPaid);

  //  What is the present value (e.g., the initial investment) of an investment that
  //  needs to total $20000 after 10 years of saving $100 every month?  Assume the
  //  interest rate is 7% (annually) compounded monthly.
  print(Finance.pv(rate: 0.07 / 12, nper: 10 * 12, pmt: -100, fv: 20000));

  //  What is the interest rate for a $8,000 loan if the loan term is 5 years and
  //  and payments are $152.50 monthly?
  print(Finance.rate(nper: 60, pmt: -152.5, pv: 8000, fv: 0));

  //  What is the Internal Rate of return of an investment that requires $15000 deposit now
  //  and then makes the following withdrawals at regular (fixed)
  //  intervals:  1500, 2500, 3500, 4500, 6000?.  Assuming the ending value is 0.
  print(Finance.irr(values: <num>[-15000, 1500, 2500, 3500, 4500, 6000]));

  //  What is the Net Present Value of an investment that requires $15000 deposit now
  //  and then makes the following withdrawals at regular (fixed)
  //  intervals:  1500, 2500, 3500, 4500, 6000?.  Assuming the ending value is 0
  //  and a discount rate of 5%.

  print(Finance.npv(
      rate: 0.05, values: <num>[-15000, 1500, 2500, 3500, 4500, 6000]));
}
